text/0000-safer-transmute.md

```diff
@@ -825,9 +692,76 @@ pub mod transmute {
             impl Sealed for () {}
             impl Sealed for NeglectAlignment {}
             impl Sealed for NeglectValidity {}
+            impl Sealed for NeglectSafety {}
             impl Sealed for (NeglectAlignment, NeglectValidity) {}
+            impl Sealed for (NeglectAlignment, NeglectSafety) {}
+            impl Sealed for (NeglectSafety, NeglectValidity) {}
+            impl Sealed for (NeglectAlignment, NeglectSafety, NeglectValidity) {}
         }
     }
+
+    /// Traits for declaring the SemVer stability of types.
+    pub mod stability {
+
+        /// Denotes that `Self`'s fields may be arbitarily initialized or
+        /// modified, regardless of their visibility. Implementing this trait
+        /// additionally denotes that you will treat any observable changes to
+        /// `Self`'s layout as breaking changes. (Unobservable changes, such as
+        /// renaming a private field, are fine.)
+        pub trait Muckable {}
+
+        /// `#[derive(Muckable)]`
+        pub macro Muckable($expr: expr) {
+            /* stub */
+        }
+
+        impl Muckable for     ! {}
+        impl Muckable for    () {}
+
+        impl Muckable for   f32 {}
+        impl Muckable for   f64 {}
+
+        impl Muckable for    i8 {}
+        impl Muckable for   i16 {}
+        impl Muckable for   i32 {}
+        impl Muckable for   i64 {}
+        impl Muckable for  i128 {}
+        impl Muckable for isize {}
+
+        impl Muckable for    u8 {}
+        impl Muckable for   u16 {}
+        impl Muckable for   u32 {}
+        impl Muckable for   u64 {}
+        impl Muckable for  u128 {}
+        impl Muckable for usize {}
```

Comment:
Shouldn't the following types also implement `Muckable`?
* `char`
* `Option<NonZero{N}>`

Reply:
`char` does not allow the full range of `u32` to be stored in it (afaict the compiler does not currently rely on this for optimizing, but it is a validity invariant so it could). If the compiler is capable of understanding its allowed range enough to implement `Muckable` for it (which would only give `char -> u32`), then shouldn't all of `NonZero{N}` also be `Muckable`? (e.g. `NonZeroU32 -> u32` seems sound).

Reply:
size_of::<char>() == size_of::<Option<char>>, because the compiler does use the niche.

Reply:
`char` and `NonZero*` and `Option` should be `Muckable`.

Muckable does *not* mean a type has no validity invariants. It means a type has no *library* validity invariants. The validity invariants of `char` and `NonZero*` are language-level.

Reply:
Weird edge case in this kind of wording, which may not mean anything in practice:

It's possible for a language-level validity invariant to move to a library validity invariant, e.g. rust-lang/reference#792. If `str` had implemented `Muckable`, because all its invariants were part of the language, would that change have required removing the impl as well?

Reply:
`char` and `bool` could also move to library-implemented types in the future, so if we stably implement `Muckable` we'd need to have the compiler somehow made aware of the exact validity requirements (like how `NonZeroU32` does).

```rust
#[lang = "bool"]
#[repr(u8)]
pub enum bool {
    false = 0,
    true = 1,
}
```

Reply:
> ```rust
> #[lang = "bool"]
> #[repr(u8)]
> pub enum bool {
>     false = 0,
>     true = 1,
> }
> ```

This is *still* an instance of language-level validity invariants.

A library-level `bool` would look something like this:
```rust
#[repr(C)] pub struct bool(u8);

const false: bool = bool(0);
const true: bool = bool(1);
```

And if you're doing anything to make the *compiler* aware of the validity requirements, you're back to language-level validity.

Reply:
> Muckable does not mean a type has no validity invariants. It means a type has no library validity invariants. The validity invariants of char and NonZero* are language-level.

I thought it means "transmutes to layout-compatible types are guaranteed to remain possible in the future". I have no idea what this means for primitive types...

I don't think the language-level validity invariant should come into play here, `Muckable` is a library concept and should be concerned with library invariants.